### PR TITLE
Port tests from `ClientRegressionWithRealNetworkTest` [API-1319]

### PR DESCRIPTION
**Pull request by emreyigit**
Tests are ported from `ClientRegressionWithRealNetworkTest `

- `TestClientConnectionBeforeServerReady` -> Client waits for cluster to connect until to be ready.
- `TestConnectionCountAfterClientReconnect` -> Client recovers when cluster not reachable, and manages active connections  accordingly.
- `TestListenersAfterClientDisconnected` -> Similiar to `TestConnectionCountAfterClientReconnect`, and it checks listeners recover.
- `TestOperationsContinueWhenClientDisconnected` -> This test is not applicable fully at the moment since reconnect modes act same, see #503  and #603. Also, I changed some parts due to lack of control capability on cluster, such as creating a key on a specific member and partition (this part tests the server side, so test is still valid for the client)

### CODE REVIEW

File: src/Hazelcast.Net/Clustering/ClusterConnections.cs [hunk 1]
```diff
@@ -330,6 +330,15 @@ private async ValueTask OnConnectionClosed(MemberConnection connection)
 
         #endregion
 
+        #region Properties
+
+        /// <summary>
+        /// Gets number of connections to members
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net/Clustering/ClusterConnections.cs:336]:
```suggestion
        /// Gets the number of open connections to members.
```

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+                .ClearAll()
+                .Configure<HConsoleLoggerProvider>().SetPrefix("LOG").SetMaxLevel()
+                .Configure().SetMinLevel().EnableTimeStamp(origin: DateTime.Now)
+                .Configure(this).SetMaxLevel().SetPrefix("TEST")
+            );
+
+        [TearDown]
+        public async Task TearDown()
+        {
+            //since tests are network related may need fresh members on cases.
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:45]: that base method does not remove members from `RcMembers` so you'll need to either fix the base method (add an `RcMembers.Clear()` of some sort, I guess...) or code a variant here that *does* clear `RcMembers`.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            await MembersOneTimeTearDown();
+        }
+
+        [Test]
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:49]:
```suggestion
        // this test validates that a client that starts connecting to a cluster that is not yet ready,
        // will eventually succeeds and connect once the cluster becomes ready.
        [Test]
```

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+            var client = await clientTask;
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(ClientState.Connected, client.State);
+            }, 10_000, 500);
+        }
+
+        [Test]
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:77]:
```suggestion
        // this test validates that a client can lost its connection to a cluster, and then reconnects,
        // and ends up with one only connection (no leak of the previous connection).
        [Test]
```

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        [Timeout(20_000)]
+        [TestCase(true, "localhost", "127.0.0.1")]
+        [TestCase(true, "127.0.0.1", "localhost")]
+        [TestCase(true, "localhost", "localhost")]
+        [TestCase(false, "localhost", "127.0.0.1")]
+        [TestCase(false, "127.0.0.1", "localhost")]
+        [TestCase(false, "localhost", "localhost")]
+        public async Task TestConnectionCountAfterClientReconnect(bool smartRouting, string memberAddress, string clientAddress)
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:85]: What are we testing exactly with `memberAddress` vs `clientAddress`?

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+            Assert.AreEqual(1, clientImp.Cluster.Connections.Count);
+            await map.PutAsync(1, 1);
+
+            await RcClient.ShutdownMemberAsync(customCluster.Id, member.Uuid);
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(ClientState.Disconnected, client.State);
+            }, 10_000, 500);
+
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:115]: Add a test for connections count, should be zero?

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            #region TearDown
+            await client.DisposeAsync();
+            await RcClient.ShutdownClusterAsync(customCluster);
+            #endregion
+
+        }
+
+
+        [Test]
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:135]:
```suggestion
        // this test validates that event handlers (listeners) that were installed, are correctly 
        // installed on a new cluster after the client has disconnected and reconnected.
        [Test]
```

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(1, client.Members.Count);
+                Assert.AreEqual(member2.Host, client.Members.First().Member.Address.Host);
+                Assert.AreEqual(member2.Port, client.Members.First().Member.Address.Port);
+
+            }, 20_000, 500);
+
+            await map.PutAsync(1, 2);
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:228]: not sure what we are testing here?

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 5 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:89]: This is ported from Java - so this is probably testing too many things including name resolution but... meh.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 8 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:234]: just testing that it works...

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+            await RcClient.ShutdownMemberAsync(customCluster.Id, member1.Uuid);
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(1, client.Members.Count);
+                Assert.AreEqual(member2.Host, client.Members.First().Member.Address.Host);
+                Assert.AreEqual(member2.Port, client.Members.First().Member.Address.Port);
+
+            }, 20_000, 500);
+
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:233]: add a comment specifying: "Java test here also toys with partitions to verify that data migrates between members - irrrelevant to us"

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 2 above)
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:45]: Fixed at 1994fec7c1ec5775d246230ddb9ab7f388640b4b

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 6 above)
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:119]: Fixed at 1994fec7c1ec5775d246230ddb9ab7f388640b4b

**Review by zpqrtbnk — APPROVED** [review]
Approved.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+            var clientTask = CreateAndStartClientAsync(options =>
+            {
+                options.Networking.SmartRouting = smartRouting;
+                options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = int.MaxValue;
+                options.Networking.ReconnectMode = ReconnectMode.ReconnectAsync;
+                options.ClusterName = RcCluster.Id;
+            });
+
+            //Let the client tries to connect before cluster is ready
+            await Task.Delay(500);
```
Comment (ihsandemir) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:67]: no such delay at Java test. if needed we should also put to Java. Otherwise lets sync both.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+            //Client will reconnect to cluster eventually.
+            member = await RcClient.StartMemberAsync(customCluster.Id);
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(ClientState.Connected, client.State);
+            }, 10_000, 500);
+
+            await map.PutAsync(1, 1);
```
Comment (ihsandemir) [inline src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:130]:
```suggestion
```

File: src/Hazelcast.Net/Clustering/ClusterConnections.cs [hunk 12]
```diff
@@ -330,6 +330,15 @@ private async ValueTask OnConnectionClosed(MemberConnection connection)
 
         #endregion
 
+        #region Properties
+
+        /// <summary>
+        /// Gets the number of open connections to members.
+        /// </summary>
+        public int Count => _connections.Count;
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net/Clustering/ClusterConnections.cs:338]: Yeap. It's internal.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        [TestCase(false)]
+        public async Task TestClientConnectionBeforeServerReady(bool smartRouting)
+        {
+            using var _ = HConsoleForTest();
+
+            var clientTask = CreateAndStartClientAsync(options =>
+            {
+                options.Networking.SmartRouting = smartRouting;
+                options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = int.MaxValue;
+                options.Networking.ReconnectMode = ReconnectMode.ReconnectAsync;
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:62]: `Async` and `Sync` modes are same at the .Net client.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 10 above)
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:67]: Java client  uses latch, we hold the client before creating a member. It's same approach here.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+                options.ClusterName = RcCluster.Id;
+            });
+
+            //Let the client tries to connect before cluster is ready
+            await Task.Delay(500);
+
+            var member = await AddMember();
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:69]: So, naming contradicts than? I'm confused what we test here.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+               options.Networking.SmartRouting = smartRouting;
+               options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = 10_000;
+               options.Networking.ReconnectMode = ReconnectMode.ReconnectAsync;
+               options.Networking.Addresses.Clear();
+               options.Networking.Addresses.Add(clientAddress + ":" + member.Port);
+               options.ClusterName = customCluster.Id;
+           });
+            #endregion
+
+            var map = await client.GetMapAsync<int, int>("TestMap_" + DateTime.UtcNow.Millisecond);
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:107]: It's just a additional check.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(1, client.Members.Count);
+                Assert.AreEqual(member2.Host, client.Members.First().Member.Address.Host);
+                Assert.AreEqual(member2.Port, client.Members.First().Member.Address.Port);
+
+            }, 20_000, 500);
+            //Java test here also toys with partitions to verify that data migrates between members - irrrelevant to us
+            //details: https://github.com/hazelcast/hazelcast-csharp-client/pull/614
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:236]: I don't see relevance here. That method checks partition synchronization between members. Am I missing something?

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                var clientImpl = (HazelcastClient)client;
+                Assert.AreEqual(1, clientImpl.Cluster.Connections.Count);
+                Assert.AreEqual(1, clientImpl.Members.Count);
+            }, 10_000, 500);
+
+            await RcClient.ShutdownMemberAsync(customCluster.Id, member.Uuid);
+            await Task.Delay(client.Options.Heartbeat.PeriodMilliseconds);
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs]: Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/614/commits/f22027eb74c23b89a057ab70a88ec45119320864

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+        [TestCase(false, "localhost", "127.0.0.1")]
+        [TestCase(false, "127.0.0.1", "localhost")]
+        [TestCase(false, "localhost", "localhost")]
+        public async Task TestListenersAfterClientDisconnected(bool smartRouting, string memberAddress, string clientAddress)
+        {
+            #region SetUp
+            using var _ = HConsoleForTest();
+            var countOfEvent = 0;
+            var memberConfig = GetMemberConfigWithAddress(memberAddress);
+            var customCluster = await RcClient.CreateClusterAsync(memberConfig);
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:159]: Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/614/commits/f22027eb74c23b89a057ab70a88ec45119320864

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+        {
+            using var _ = HConsoleForTest();
+            #region SetUp
+            var memberConfig = GetMemberConfigWithAddress(memberAddress);
+            var customCluster = await RcClient.CreateClusterAsync(memberConfig);
+            var member = await RcClient.StartMemberAsync(customCluster.Id);
+            var client = await CreateAndStartClientAsync(options =>
+           {
+               options.Networking.SmartRouting = smartRouting;
+               options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = 10_000;
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs]: Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/614/commits/f22027eb74c23b89a057ab70a88ec45119320864

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+            {
+                Assert.AreEqual(ClientState.Connected, client.State);
+            }, 10_000, 500);
+        }
+
+        // this test validates that a client can lost its connection to a cluster, and then reconnects,
+        // and ends up with one only connection (no leak of the previous connection).
+        [Test]
+        [Timeout(20_000)]
+        [TestCase(true, "localhost", "127.0.0.1")]
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:83]: Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/614/commits/f22027eb74c23b89a057ab70a88ec45119320864

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (ihsandemir) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:62]: I mean why don't you also have the one with `ReconnectMode.ReconnectSync`

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 10 above)
Comment (ihsandemir) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:67]: Time dependent tests are not reliable especially when using github action machines. Why don@t you have latch usage in .NET?

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ihsandemir) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:69]: To match the same as Java test, line 69 where you add member should be before you start client, right?

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (ihsandemir) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:239]: Please match Java code otherwise send PR to Java to change it if not relevant. I think that it has a logic that uses a key owned by member2 which is not shutdown and yes, it may not even be migrating since it will be backup promoting to primary.

[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs:241]: Opened a ticket to discuss to change it as well on Java side not to block the PR. Still, I don't see relevance with the client tests, let's discuss it.

File: src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs]: Because they behave same but I added anyway, incase of we can change their implementation in the future. Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/614/commits/93423d885cc9f88fba8896ebe4b1990e801f3a45

**Review by zpqrtbnk — APPROVED** [review]
Approved

**Review by yuce — APPROVED** [review]
